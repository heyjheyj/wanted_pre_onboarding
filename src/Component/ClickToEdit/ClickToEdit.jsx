import React, { useState } from "react";
import styles from "./ClickToEdit.module.css";

const ClickToEdit = ({ initEdit }) => {
  const [name, setName] = useState(initEdit.name);
  const [age, setAge] = useState(initEdit.age);
  const [currentFocus, setCurrentFocus] = useState("");

  const changeName = e => {
    setName(e.target.value);
  };

  const changeAge = e => {
    setAge(e.target.value);
  };

  return (
    <div className={styles.container}>
      <h3 className={styles.title}>ClickToEdit</h3>
      <div className={styles.name}>
        <label htmlFor="username">
          이름
          <input
            className={
              currentFocus === "name"
                ? `${styles.nameinput}`
                : `${styles.input}`
            }
            type="text"
            id="username"
            data-testid="test-nameinput"
            defaultValue={name}
            onFocus={() => setCurrentFocus("name")}
            onBlur={changeName}
          />
        </label>
      </div>
      <div className={styles.age}>
        <label htmlFor="userage">
          나이
          <input
            className={
              currentFocus === "age" ? `${styles.ageinput}` : `${styles.input}`
            }
            type="text"
            id="userage"
            data-testid="test-ageinput"
            defaultValue={age}
            onFocus={() => setCurrentFocus("age")}
            onBlur={changeAge}
          />
        </label>
      </div>
      <span>
        이름 {name} 나이 {age}
      </span>
    </div>
  );
};

export default ClickToEdit;
