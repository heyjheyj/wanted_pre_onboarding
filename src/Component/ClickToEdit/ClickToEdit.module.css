.container {
    margin-top: 10px;
    width: 80%;
    height: 200px;
    border: 1px solid grey;
    display: flex;
    flex-direction: column;
    position: relative;
    align-items: center;
    justify-content: space-evenly;
    border-radius: 10px;
}

.title {
    position: absolute;
    top: 5px;
    left: 5px;
}

.input, .nameinput, .ageinput {
    margin-left: 5px;
    text-align: center;
    font-size: 15px;
}

.nameinput:focus-within, .ageinput:focus-within {
    border: 2px solid #431cc6;
    border-radius: 2px;
}
