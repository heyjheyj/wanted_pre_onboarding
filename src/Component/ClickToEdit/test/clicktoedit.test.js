import React from "react";
import ClickToEdit from "../ClickToEdit";
import TestRenderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { render, screen } from "@testing-library/react";

describe("ClickToEdit", () => {
  let initEdit = { name: "김코딩", age: "20" };

  it("renders", () => {
    const component = TestRenderer.create(<ClickToEdit initEdit={initEdit} />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  describe("Edit text by Typing on input", () => {
    let nameinput;
    let ageinput;

    beforeEach(() => {
      render(<ClickToEdit initEdit={initEdit} />);
      nameinput = screen.getByTestId("test-nameinput");
      ageinput = screen.getByTestId("test-ageinput");
    });

    it("focus on nameinput when is clicked and changes color of border from 'default' to 'focused color'", () => {
      nameinput.focus();

      expect(document.activeElement).toBe(nameinput);
      expect(nameinput.className).toBe("nameinput");
    });

    it("blur of nameinput when outside of inputbox is clicked and changes color of border from 'focused color' to 'default'", () => {
      nameinput.blur();

      expect(nameinput.value).toBe("김코딩");
      expect(nameinput.className).toBe("input");
    });

    it("focus on ageinput when is clicked and changes color of border from 'default' to 'focused color' ", () => {
      ageinput.focus();

      expect(document.activeElement).toBe(ageinput);
      expect(ageinput.className).toBe("ageinput");
    });

    it("blur of ageinput when outside of inputbox is clicked and change color of border from 'focused color' to 'default'", () => {
      ageinput.blur();

      expect(ageinput.value).toBe("20");
      expect(ageinput.className).toBe("input");
    });

    it("changes a nameinput value when blur event fires", () => {
      userEvent.clear(nameinput);
      userEvent.type(nameinput, "최해커");
      nameinput.blur();

      expect(nameinput.value).toBe("최해커");
    });

    it("changes a ageinput value when blur event fires", () => {
      userEvent.clear(ageinput);
      userEvent.type(ageinput, "30");
      ageinput.blur();

      expect(ageinput.value).toBe("30");
    });
  });
});
