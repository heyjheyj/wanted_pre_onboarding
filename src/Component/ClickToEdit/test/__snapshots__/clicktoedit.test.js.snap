// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickToEdit renders 1`] = `
<div
  className="container"
>
  <h3
    className="title"
  >
    ClickToEdit
  </h3>
  <div
    className="name"
  >
    <label
      htmlFor="username"
    >
      이름
      <input
        className="input"
        data-testid="test-nameinput"
        defaultValue="김코딩"
        id="username"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
      />
    </label>
  </div>
  <div
    className="age"
  >
    <label
      htmlFor="userage"
    >
      나이
      <input
        className="input"
        data-testid="test-ageinput"
        defaultValue="20"
        id="userage"
        onBlur={[Function]}
        onFocus={[Function]}
        type="text"
      />
    </label>
  </div>
  <span>
    이름 
    김코딩
     나이 
    20
  </span>
</div>
`;
